cpp/src/text/vocabulary_tokenize.cu

```diff
@@ -162,6 +173,122 @@ std::unique_ptr<tokenize_vocabulary> load_vocabulary(cudf::strings_column_view c
+  int8_t* d_results;
+
+  __device__ void operator()(cudf::size_type idx) const
+  {
+    auto const ptr = d_chars + idx;
+    if (cudf::strings::detail::is_utf8_continuation_char(*ptr)) { return; }
+    cudf::char_utf8 chr = 0;
+    auto ch_size        = cudf::strings::detail::to_char_utf8(ptr, chr);
+    auto const output   = is_delimiter(d_delimiter, chr);
+    while (ch_size > 0) {
+      d_results[idx++] = output;
+      --ch_size;
+    }
+  }
+};
+
+__global__ void token_counts_fn(cudf::column_device_view const d_strings,
+                                cudf::string_view const d_delimiter,
+                                cudf::size_type* d_counts,
+                                int8_t* d_results)
+{
+  // string per warp
+  auto const idx = static_cast<std::size_t>(threadIdx.x + blockIdx.x * blockDim.x);
+  if (idx >= (static_cast<std::size_t>(d_strings.size()) *
+              static_cast<std::size_t>(cudf::detail::warp_size))) {
+    return;
+  }
+  auto const str_idx  = static_cast<cudf::size_type>(idx / cudf::detail::warp_size);
+  auto const lane_idx = static_cast<cudf::size_type>(idx % cudf::detail::warp_size);
+
+  if (d_strings.is_null(str_idx)) {
+    d_counts[str_idx] = 0;
+    return;
+  }
+  auto const d_str = d_strings.element<cudf::string_view>(str_idx);
+  if (d_str.empty()) {
+    d_counts[str_idx] = 0;
+    return;
+  }
+
+  auto const offsets =
+    d_strings.child(cudf::strings_column_view::offsets_column_index).data<cudf::size_type>();
+  auto const offset = offsets[str_idx + d_strings.offset()] - offsets[d_strings.offset()];
+  auto const chars_begin =
+    d_strings.child(cudf::strings_column_view::chars_column_index).data<char>() +
+    offsets[d_strings.offset()];
+
+  auto const begin        = d_str.data();
+  auto const end          = begin + d_str.size_bytes();
+  auto const d_output     = d_results + offset;
+  auto const d_output_end = d_output + d_str.size_bytes();
+
+  using warp_reduce = cub::WarpReduce<cudf::size_type>;
+  __shared__ typename warp_reduce::TempStorage warp_storage;
+
+  cudf::size_type count = 0;
+  if (lane_idx == 0) {
+    cudf::char_utf8 chr = 0;
+    auto ch_size        = cudf::strings::detail::to_char_utf8(begin, chr);
+    auto output         = 1;
+    if (begin > chars_begin) {
+      auto ptr = begin - 1;
+      while (ptr > chars_begin && cudf::strings::detail::is_utf8_continuation_char(*ptr)) {
+        --ptr;
+      }
+      cudf::strings::detail::to_char_utf8(ptr, chr);
+      output = !is_delimiter(d_delimiter, chr);
+    }
+    auto ptr = d_output;
+    while (ch_size > 0) {
+      *ptr++ = output;
+      --ch_size;
+    }
+    count = ((begin + ch_size) == end);
+  }
+  __syncwarp();
+
+  for (auto itr = d_output + lane_idx + 1; itr < d_output_end; itr += cudf::detail::warp_size) {
+    auto index = static_cast<cudf::size_type>(thrust::distance(d_output, itr));
+    if (*itr) {
+      count += !(*(itr - 1));
+    } else if (itr + 1 == d_output_end) {
+      ++count;
+    }
```

Comment:
Is this equivalent? It seems like there are a few cases that add 1, but they're spelled differently.
```suggestion
    if ((*itr && !(*(itr - 1))) || (itr + 1 == d_output_end)) {
      ++count;
    }
```

Reply:
I initially thought the original version was more readable.  I'm now leaning towards preferring @bdice's version.

Reply:
See the current implementation. I added a comment here as well that I helps.

Reply:
Forgive me, but I think this one-liner is too terse now.

I'll go along with it, if the consensus is that this is best.  But I don't think this would be readable to me, had I not seen the evolution from the original version.

Reply:
The key outcome I want is to have only one way to get to "count++". You could also achieve that by declaring two named bools and OR-ing them, rather than putting both conditions directly into the if.

Reply:
Surely the comment above the line should alleviate any concerns?

Reply:
Sorry. I was replying to Mithun on the general topic of terseness / one-liners. I like the current state of the code, the comment is helpful here.

Reply:
> The key outcome I want is to have only one way to get to "count++". 

Thanks, that helped me grok this better.

Reply:
Turns out this is not equivalent to the original if-else.
Here is the desired logic table:
```
a = *itr
b = !(*(itr-1))
c = (itr + 1) == d_output_end
r = count
0 = false, 1 =true

 a  b  c    r
 0  0  0    -
 0  0  1   +1
 0  1  0    -
 0  1  1   +1
 1  0  0    -
 1  0  1    -     <<<< new logic incorrectly adds 1 to count here
 1  1  0   +1
 1  1  1   +1

if( (a && b) || c) {r++}  is not the same as  if(a) {r+=b} else if(c) {r++}
```
(Sorry, the table may be confusing since `!*(itr-1) == true` when the value is actually false).
Anyway, the new code is causing failures that my tests are not catching so I need to revert this change and add some additional tests.

Reply:
Yes, I agree. I see why -- you need to skip checking `c` if `a` is true. I tried to work that truth table out in my head while reviewing but made an error in that case. Apologies for the incorrect suggestion.